cbmgmtx/mgmt.go

```diff
@@ -1016,3 +1018,164 @@ func (h Management) FlushBucket(
 	_ = resp.Body.Close()
 	return nil
 }
+
+// AuthDomain specifies the user domain of a specific user
+type AuthDomain string
+
+const (
+	// AuthDomainAdmin specifies users that are locally stored as Administrators.
+	AuthDomainAdmin AuthDomain = "admin"
+
+	// AuthDomainLocal specifies users that are locally stored in Couchbase.
+	AuthDomainLocal AuthDomain = "local"
+
+	// AuthDomainExternal specifies users that are externally stored
+	// (in LDAP for instance).
+	AuthDomainExternal AuthDomain = "external"
+)
+
+type OriginJson struct {
+	Type string `json:"type"`
+	Name string `json:"name"`
+}
+
+type RoleJson struct {
+	RoleName       string `json:"role"`
+	BucketName     string `json:"bucket_name"`
+	ScopeName      string `json:"scope_name"`
+	CollectionName string `json:"collection_name"`
+}
+
+type RoleWithOriginsJson struct {
+	RoleJson
+
+	Origins []OriginJson
+}
+
+type UserJson struct {
+	ID              string                `json:"id"`
+	Name            string                `json:"name"`
+	Roles           []RoleWithOriginsJson `json:"roles"`
+	Groups          []string              `json:"groups"`
+	Domain          AuthDomain            `json:"domain"`
+	ExternalGroups  []string              `json:"external_groups"`
+	PasswordChanged time.Time             `json:"password_change_date"`
+}
+
+type GetAllUsersOptions struct {
+	OnBehalfOf *cbhttpx.OnBehalfOfInfo
+}
+
+func (h Management) GetAllUsers(
+	ctx context.Context,
+	opts *GetAllUsersOptions,
+) ([]*UserJson, error) {
+	resp, err := h.Execute(
+		ctx,
+		"GET",
+		"/settings/rbac/users",
+		"", opts.OnBehalfOf, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != 200 {
+		return nil, h.DecodeCommonError(resp)
+	}
+
+	usersData, err := cbhttpx.ReadAsJsonAndClose[[]*UserJson](resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return usersData, nil
+}
+
+type UpsertUserOptions struct {
+	Domain      AuthDomain
+	Username    string
+	DisplayName string
+	Password    string
+	Roles       []string
+	Groups      []string
+	OnBehalfOf  *cbhttpx.OnBehalfOfInfo
+}
+
+func (h Management) UpsertUser(
+	ctx context.Context,
+	opts *UpsertUserOptions,
+) error {
+	if opts.Username == "" {
+		return errors.New("must specify username when upserting a user")
+	}
+
+	if opts.Domain == "" {
+		opts.Domain = "local"
+	}
+
+	posts := url.Values{}
+
+	if opts.DisplayName != "" {
+		posts.Add("name", opts.DisplayName)
+	}
+	if opts.Password != "" {
```

Comment:
Is it valid to create a user without a password?

Reply:
I'm not entirely sure, but I didn't think there was much harm in avoiding sending the field if the field is blank anyways.  I suspect you're underlying assumption is probably correct though, and the server would reject the request.